xds/src/main/java/io/grpc/xds/CdsLoadBalancer.java

```diff
@@ -350,16 +230,33 @@ public void onResourceDoesNotExist(String resourceName) {
 
     @Override
     public void onError(Status error) {
-      logger.log(
-          XdsLogLevel.WARNING,
-          "Received error from xDS client {0}: {1}: {2}",
-          xdsClient,
-          error.getCode(),
-          error.getDescription());
-      if (edsBalancer == null) {
+      logger.log(XdsLogLevel.WARNING, "Received error from xDS client {0}: {1}", error);
+      if (lbHelper.lb == null) {
         helper.updateBalancingState(TRANSIENT_FAILURE, new ErrorPicker(error));
       }
     }
-  }
 
+    void start() {
+      xdsClient.watchClusterData(clusterName, this);
+      logger.log(XdsLogLevel.INFO,
+          "Started watcher for cluster {0} with xDS client {1}", clusterName, xdsClient);
+    }
+
+    void shutdown() {
+      xdsClient.cancelClusterDataWatch(clusterName, this);
+      logger.log(XdsLogLevel.INFO,
+          "Cancelled watcher for cluster {0} with xDS client {1}", clusterName, xdsClient);
+      if (lbHelper.lb != null) {
+        lbHelper.lb.shutdown();
+      }
+    }
+
+    void propagateError(Status error) {
+      if (lbHelper.lb != null) {
+        lbHelper.lb.handleNameResolutionError(error);
+      } else {
+        lbHelper.updateBalancingState(TRANSIENT_FAILURE, new ErrorPicker(error));
```

Comment:
This can be `helper.updateBalancingState(TRANSIENT_FAILURE, new ErrorPicker(error))` because `lbHelper` is only for `lb`. If `lb` is `null`, you should use `helper` directly.

Reply:
Yay, reasonable. Changed.